File: backend/src/modules/permissions/permissions.controller.ts
import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  Query,
  ParseIntPipe,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse } from '@nestjs/swagger';
import { PermissionsService } from './permissions.service';
import { CreateAddressDto } from './dto/create-address.dto';
import { CreateTrustedDto } from './dto/create-trusted.dto';
import { PaginationDto } from '../../common/dto/pagination.dto';
import { ApiResponseDto } from '../../common/dto/api-response.dto';

@ApiTags('权限管理')
@Controller('api/permissions')
export class PermissionsController {
  constructor(private readonly permissionsService: PermissionsService) {}

  // ========== Address 接口 ==========

  @Post('address')
  @ApiOperation({ summary: '创建地址白名单' })
  async createAddress(@Body() createDto: CreateAddressDto): Promise<ApiResponseDto> {
    const address = await this.permissionsService.createAddress(createDto);
    return ApiResponseDto.success(address, '创建成功');
  }

  @Get('address')
  @ApiOperation({ summary: '获取地址白名单列表' })
  async findAllAddresses(@Query() paginationDto: PaginationDto): Promise<ApiResponseDto> {
    const result = await this.permissionsService.findAllAddresses(paginationDto);
    return ApiResponseDto.success(result);
  }

  @Get('address/:id')
  @ApiOperation({ summary: '获取地址详情' })
  async findOneAddress(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    const address = await this.permissionsService.findOneAddress(id);
    return ApiResponseDto.success(address);
  }

  @Patch('address/:id')
  @ApiOperation({ summary: '更新地址' })
  async updateAddress(
    @Param('id', ParseIntPipe) id: number,
    @Body() updateDto: Partial<CreateAddressDto>,
  ): Promise<ApiResponseDto> {
    const address = await this.permissionsService.updateAddress(id, updateDto);
    return ApiResponseDto.success(address, '更新成功');
  }

  @Delete('address/:id')
  @ApiOperation({ summary: '删除地址' })
  async removeAddress(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    await this.permissionsService.removeAddress(id);
    return ApiResponseDto.success(null, '删除成功');
  }

  // ========== Trusted 接口 ==========

  @Post('trusted')
  @ApiOperation({ summary: '创建信任地址' })
  async createTrusted(@Body() createDto: CreateTrustedDto): Promise<ApiResponseDto> {
    const trusted = await this.permissionsService.createTrusted(createDto);
    return ApiResponseDto.success(trusted, '创建成功');
  }

  @Get('trusted')
  @ApiOperation({ summary: '获取信任地址列表' })
  async findAllTrusted(@Query() paginationDto: PaginationDto): Promise<ApiResponseDto> {
    const result = await this.permissionsService.findAllTrusted(paginationDto);
    return ApiResponseDto.success(result);
  }

  @Get('trusted/:id')
  @ApiOperation({ summary: '获取信任地址详情' })
  async findOneTrusted(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    const trusted = await this.permissionsService.findOneTrusted(id);
    return ApiResponseDto.success(trusted);
  }

  @Patch('trusted/:id')
  @ApiOperation({ summary: '更新信任地址' })
  async updateTrusted(
    @Param('id', ParseIntPipe) id: number,
    @Body() updateDto: Partial<CreateTrustedDto>,
  ): Promise<ApiResponseDto> {
    const trusted = await this.permissionsService.updateTrusted(id, updateDto);
    return ApiResponseDto.success(trusted, '更新成功');
  }

  @Delete('trusted/:id')
  @ApiOperation({ summary: '删除信任地址' })
  async removeTrusted(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    await this.permissionsService.removeTrusted(id);
    return ApiResponseDto.success(null, '删除成功');
  }

  // ========== 通用接口 ==========

  @Post('reload')
  @ApiOperation({ summary: '重载权限配置' })
  async reload(): Promise<ApiResponseDto> {
    await this.permissionsService.reload();
    return ApiResponseDto.success(null, '重载成功');
  }

  @Get('stats')
  @ApiOperation({ summary: '获取权限统计信息' })
  async getStats(): Promise<ApiResponseDto> {
    const stats = await this.permissionsService.getStats();
    return ApiResponseDto.success(stats);
  }
}

File: kam-admin-module-user/src/main/java/io/kamailio/admin/modules/permissions/controller/PermissionsController.java
package io.kamailio.admin.modules.permissions.controller;

import io.kamailio.admin.common.dto.ApiResponse;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.permissions.dto.CreateAddressDto;
import io.kamailio.admin.modules.permissions.dto.CreateTrustedDto;
import io.kamailio.admin.modules.permissions.entity.Address;
import io.kamailio.admin.modules.permissions.entity.Trusted;
import io.kamailio.admin.modules.permissions.service.PermissionsService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@Tag(name = "权限管理")
@RestController
@RequestMapping("/api/permissions")
public class PermissionsController {
    private final PermissionsService service;

    public PermissionsController(PermissionsService service) { this.service = service; }

    @PostMapping("/address")
    @Operation(summary = "创建地址白名单")
    public ResponseEntity<ApiResponse<Address>> createAddress(@Valid @RequestBody CreateAddressDto dto) {
        return ResponseEntity.ok(ApiResponse.success(service.createAddress(dto), "创建成功"));
    }

    @GetMapping("/address")
    @Operation(summary = "获取地址白名单列表")
    public ResponseEntity<ApiResponse<PaginatedResult<Address>>> findAllAddresses(PaginationDto dto) {
        return ResponseEntity.ok(ApiResponse.success(service.findAllAddresses(dto)));
    }

    @GetMapping("/address/{id}")
    @Operation(summary = "获取地址详情")
    public ResponseEntity<ApiResponse<Address>> findOneAddress(@PathVariable Integer id) {
        return ResponseEntity.ok(ApiResponse.success(service.findOneAddress(id)));
    }

    @PatchMapping("/address/{id}")
    @Operation(summary = "更新地址")
    public ResponseEntity<ApiResponse<Address>> updateAddress(@PathVariable Integer id, @RequestBody CreateAddressDto dto) {
        return ResponseEntity.ok(ApiResponse.success(service.updateAddress(id, dto), "更新成功"));
    }

    @DeleteMapping("/address/{id}")
    @Operation(summary = "删除地址")
    public ResponseEntity<ApiResponse<Void>> removeAddress(@PathVariable Integer id) {
        service.removeAddress(id);
        return ResponseEntity.ok(ApiResponse.success(null, "删除成功"));
    }

    @PostMapping("/trusted")
    @Operation(summary = "创建信任地址")
    public ResponseEntity<ApiResponse<Trusted>> createTrusted(@Valid @RequestBody CreateTrustedDto dto) {
        return ResponseEntity.ok(ApiResponse.success(service.createTrusted(dto), "创建成功"));
    }

    @GetMapping("/trusted")
    @Operation(summary = "获取信任地址列表")
    public ResponseEntity<ApiResponse<PaginatedResult<Trusted>>> findAllTrusted(PaginationDto dto) {
        return ResponseEntity.ok(ApiResponse.success(service.findAllTrusted(dto)));
    }

    @GetMapping("/trusted/{id}")
    @Operation(summary = "获取信任地址详情")
    public ResponseEntity<ApiResponse<Trusted>> findOneTrusted(@PathVariable Integer id) {
        return ResponseEntity.ok(ApiResponse.success(service.findOneTrusted(id)));
    }

    @PatchMapping("/trusted/{id}")
    @Operation(summary = "更新信任地址")
    public ResponseEntity<ApiResponse<Trusted>> updateTrusted(@PathVariable Integer id, @RequestBody CreateTrustedDto dto) {
        return ResponseEntity.ok(ApiResponse.success(service.updateTrusted(id, dto), "更新成功"));
    }

    @DeleteMapping("/trusted/{id}")
    @Operation(summary = "删除信任地址")
    public ResponseEntity<ApiResponse<Void>> removeTrusted(@PathVariable Integer id) {
        service.removeTrusted(id);
        return ResponseEntity.ok(ApiResponse.success(null, "删除成功"));
    }

    @PostMapping("/reload")
    @Operation(summary = "重载权限配置")
    public ResponseEntity<ApiResponse<Void>> reload() {
        service.reload();
        return ResponseEntity.ok(ApiResponse.success(null, "重载成功"));
    }

    @GetMapping("/stats")
    @Operation(summary = "获取权限统计信息")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getStats() {
        return ResponseEntity.ok(ApiResponse.success(service.getStats()));
    }

    @PostMapping("/reload-trusted")
    @Operation(summary = "重载信任源表")
    public ResponseEntity<ApiResponse<Void>> reloadTrusted() {
        service.reloadTrusted();
        return ResponseEntity.ok(ApiResponse.success(null, "重载成功"));
    }

    @GetMapping("/address-dump")
    @Operation(summary = "导出内存地址表")
    public ResponseEntity<ApiResponse<Object>> addressDump() {
        return ResponseEntity.ok(ApiResponse.success(service.addressDump()));
    }

    @GetMapping("/subnet-dump")
    @Operation(summary = "导出内存子网表")
    public ResponseEntity<ApiResponse<Object>> subnetDump() {
        return ResponseEntity.ok(ApiResponse.success(service.subnetDump()));
    }
}
